### ISSUE TITLE
Ability to recover NFT Plot rewards in the machinaris container itself or automatically.

### ISSUE THREAD

**Issue by ApfelBirneKreis** [guydavis/machinaris#354]
**Describe the request**
Ability to recover NFT Plot rewards in the machinaris container itself or automatically.

On "All the Blocks" there is the recovery tool for recovering Plot rewards on NFT plots. I would love to have the abillity to do this in the machinaris container itself. The Flora devs are providing this tool here (which All the Blocks is also using):

[https://github.com/Flora-Network/flora-dev-cli](url) 

To use the tool one must provide the Launcher ID and the pool contract address in a preloaded config.

I hoped that it would be possible to integrate or run along with the Machinaris main container since it needs python and git as Prerequisites. Just to automate things. it even provides the ability to loop at custom timings.

I thought it could be implemented with a button or just the ability to enable it in the machinaris config. 

configs in the tool look like this and can be used for every fork that uses NFT chia plots (k32+)

```
# Set env var to blockchain path.
export FD_CLI_BC_DB_PATH=$HOME/.flora/mainnet/db/blockchain_v1_mainnet.sqlite

# Set env var to wallet path.
# This must be the wallet that is associated with mnemonic from which NFT plot was created. (Usually your hot wallet)
# Replace <fingerprint> with your wallet fingerprint found at below path or by using "chia wallet show"
export FD_CLI_WT_DB_PATH=$HOME/.flora/mainnet/wallet/db/blockchain_wallet_v1_mainnet_<fingerprint>.sqlite

# Set env var to launcher id of NFT plot. Replace the below ID with output of "Launcher ID:" 
# Launcher ID: can be obtained using "chia plotnft show"
# Execute above command in Chia, as those values are the original NFT contract details, which do not exist in the forks
export LAUNCHER_HASH=aaa0cbae497933a6c029a3819759fe148829dfde0316cb0512ccad23edce6aaa

# Set env var to pool_contract_address. 
# Pool contract address: can be obtained using "chia plotnft show"
# Execute above command in Chia, as those values are the original NFT contract details, which do not exist in the forks
export POOL_CONTRACT_ADDRESS=xch13rht0xz4tpdqfq08e3dk20kewg9cjj3pw0wwjf7vay8whlxn7ppqapeqhz

fd-cli nft-recover \
  -l "$LAUNCHER_HASH" \
  -p "$POOL_CONTRACT_ADDRESS" \
  -nh 127.0.0.1 \
  -np 18755 \
  -ct $HOME/.flora/mainnet/config/ssl/full_node/private_full_node.crt \
  -ck $HOME/.flora/mainnet/config/ssl/full_node/private_full_node.key
  
# All coins that were mined +7 days ago WITH NFT PLOT should be spendable soon via wallet.
```

The examples of Recovering coins on Flora blockchain:

One time recovery of coins in Flora blockchain:

`./flora_recovery.ps1 -POOL_CONTRACT_ADDRESS [your_pool_contract_address] -LAUNCHER_HASH [your_launcher_id]`

One time recovery of coins in Flora blockchain when multiple wallet are available:

`./flora_recovery.ps1 -POOL_CONTRACT_ADDRESS [your_pool_contract_address] -LAUNCHER_HASH [your_launcher_id] -fingerprint [wallet_fingerprint]`

One time recovery of coins in Flora and Silicoin blockchains:

`./flora_recovery.ps1 -POOL_CONTRACT_ADDRESS [your_pool_contract_address] -LAUNCHER_HASH [your_launcher_id] -blockchains flora, silicoin`

Recovery of coins in Flora and Silicoin blockchains in a loop, executing the recovery every 24 hours:

`./flora_recovery.ps1 -POOL_CONTRACT_ADDRESS [your_pool_contract_address] -LAUNCHER_HASH [your_launcher_id] -blockchains flora, silicoin -sleep 24`

**Comment by guydavis** [comment]
Should be available in v0.6.2 now.  Log output to /root/.chia/machinaris/logs/fd-cli.log daily, in each container executing it (not Chia and not Chives).

**Comment by ApfelBirneKreis** [comment]
I get this error in my log on HDD coin.

```
An error occurred while sending the recovery transaction.
Traceback (most recent call last):
  File "/usr/local/bin/fd-cli", line 33, in <module>
    sys.exit(load_entry_point('fd-cli', 'console_scripts', 'fd-cli')())
  File "/fd-cli/fd_cli/fd_cli.py", line 220, in main
    fd_cli()
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.9/dist-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/fd-cli/fd_cli/fd_cli.py", line 193, in fd_cli_nft_recover
    fd_cli_cmd_nft_recover(
  File "/fd-cli/fd_cli/fd_cli_cmd_nft_recover.py", line 211, in fd_cli_cmd_nft_recover
    fd_cli_print_raw(exception, pre=pre)
  File "/fd-cli/fd_cli/fd_cli_print.py", line 15, in fd_cli_print_raw
    print(f'{" " * pre * 4}{value:{fill}s}')
TypeError: unsupported format string passed to HTTPError.__format__
```

**Comment by guydavis** [comment]
Re-opening as another Unraid user [reported a similar error](https://forums.unraid.net/topic/108896-support-machinaris-chia-cryptocurrency-farming-plotman-plotting-unraid-webui/?do=findComment&comment=1049196) from the script.  Seems to be tied to the inputs to the script, however I haven't been able to replicate this myself.

**Comment by ApfelBirneKreis** [comment]
is it possible to try and stike the pool address part in the next update? i have a feeling that is the issue.

**Comment by ApfelBirneKreis** [comment]
or is there a way to edit this myself?

**Comment by guydavis** [comment]
Finally got a chance to try the command without the `-p` pool_contract_address:

![image](https://user-images.githubusercontent.com/12936/139971469-88486121-b64d-4125-943e-ffa1bebbe390.png)

Here's a redacted version of what I tried:

```
export FD_CLI_BC_DB_PATH=/root/.hddcoin/mainnet/db/blockchain_v1_mainnet.sqlite
export FD_CLI_WT_DB_PATH=/root/.hddcoin/mainnet/wallet/db/blockchain_wallet_v1_mainnet_123456789.sqlite
/usr/local/bin/fd-cli nft-recover -l REDACTED -nh 127.0.0.1 -np 28559 -ct /root/.hddcoin/mainnet/config/ssl/full_node/private_full_node.crt -ck /root/.hddcoin/mainnet/config/ssl/full_node/private_full_node.key 
```

**Comment by cjizzle4487** [comment]
Spoke to Guy on Discord. Currently the farmer node rpc port is being passed and I believe it needs to be the full node port.  He said he would push the full node ports in the next updates to the containers.

**Comment by codegrunt** [comment]
it seems to be the RPC port that you need to have it wired into - for flora that is 18755

**Comment by guydavis** [comment]
Hi all, this hasn't fallen off my radar.  Still working to fix a dumb-typo I allowed to slide thru in v0.6.3.  Corrected in latest `:develop` stream (pre v0.6.4).  Right now I am monitoring all my fork full-nodes, watching the fd-cli.log, hourly as I've sped it up from daily for testing purposes.

![20211108_135736](https://user-images.githubusercontent.com/12936/140817159-05cc8c45-840c-4312-958f-92f7a3482893.png)

**Comment by guydavis** [comment]
This fix has now been released as v0.6.4.  Let me know if the problem recurs in this latest release and I'll re-open the issue.  Thanks again for the report.
